File: tests/js/keymap.js
.import "../../ui/js/Keymap.js" as Keymap

// Default is the one preset, and its listing table is the arrows, the Ctrl chords and a handful of
// named keys. Everything this file asserts is either in that set or is asserted to be gone: a key
// that quietly stops working is the failure mode a keymap has, and nothing else notices it.
function run(check) {
    var none = Qt.NoModifier, ctrl = Qt.ControlModifier, shift = Qt.ShiftModifier
    var alt = Qt.AltModifier, meta = Qt.MetaModifier
    function key(name, text, mods, expected, context, frontend) {
        check("default " + (context || "listing") + " " + (frontend || "gui") + " " + name + " " + text + " " + mods,
              Keymap.lookupFor("default", Qt["Key_" + name] || 0, text, mods, context, frontend), expected)
    }

    check("default is the only preset", Keymap.PRESETS.join(","), "default")

    // The whole listing table, one row per key it binds.
    key("Down", "", none, "cursorDown")
    key("Up", "", none, "cursorUp")
    key("Left", "", none, "parent")
    key("Right", "", none, "pageForward")
    key("Return", "", none, "open")
    key("Enter", "", none, "open")
    // The 2026-09-18 ruling: Space marks the row and Right is what opens the preview, so the
    // listing's own Space is the selection and only the preview contexts still read it as the card.
    key("Space", " ", none, "toggleSelect")
    key("A", "", ctrl, "selectAll")
    key("Tab", "", none, "focusNext")
    key("Escape", "", none, "escape")
    key("Delete", "", none, "trash")
    key("F2", "", none, "rename")
    // F5 re-reads the directory on screen, the key every file manager on this desktop answers with.
    key("F5", "", none, "refresh")
    key("C", "", ctrl, "copy")
    key("V", "", ctrl, "paste")
    key("X", "", ctrl, "cut")
    key("Z", "", ctrl, "undo")
    key("F", "", ctrl, "search")
    key("F", "", ctrl | shift, "searchDeep")
    key("B", "", ctrl, "copypath")
    key("G", "", ctrl, "sidebar")
    key("T", "", ctrl, "toggleDual")
    // The 2026-09-18 trio. New Folder moved off Ctrl+Shift+N, so the shifted chord must now
    // resolve to nothing: a stale binding that still answered would shadow the ctrlshift fallthrough.
    key("N", "", ctrl, "newFolder")
    key("N", "", ctrl | shift, "")
    key("M", "", ctrl, "newFile")
    key("H", "", ctrl, "toggleHidden")
    key("J", "", ctrl, "sortName")
    key("K", "", ctrl, "sortSize")
    key("L", "", ctrl, "sortDate")
    // Ctrl+M and the bare m are different rows, and the unmodified one still opens the menu.
    key("M", "m", none, "menu")
    key("Menu", "", none, "menu")
    key("F10", "", shift, "menu")
    key("Question", "?", shift, "keymapSheet")
    key("Comma", ",", none, "settings")
    // The split, both ways on the one chord; the chrome's fourth mode had no key before it.
    key("T", "", ctrl, "toggleDual")

    // And the keys that left with the vim spellings, the view chords and the tab chords. A sample
    // wide enough that a row creeping back in is caught rather than a spot check of three.
    var goneLetters = [["H", "h"], ["J", "j"], ["K", "k"], ["L", "l"], ["Y", "y"], ["X", "x"],
                       ["P", "p"], ["R", "r"], ["Z", "z"], ["V", "v"], ["S", "s"], ["F", "f"],
                       ["O", "o"], ["G", "g"], ["T", "t"], ["W", "w"], ["A", "a"], ["E", "e"],
                       ["D", "d"], ["Period", "."], ["Slash", "/"], ["Minus", "-"]]
    for (var g = 0; g < goneLetters.length; g++)
        key(goneLetters[g][0], goneLetters[g][1], none, "")
    key("H", "H", shift, "")
    key("L", "L", shift, "")
    key("G", "G", shift, "")
    key("S", "S", shift, "")
    key("Colon", ":", shift, "")
    key("Plus", "+", shift, "")
    // Ctrl+T is not here: it lost the terminal and now carries the split, checked in the table
    // above. Nor are Ctrl+N, Ctrl+M and Ctrl+H, which the 2026-09-18 trio took for New Folder,
    // New File and the hidden toggle; all three are in that table too.
    // Ctrl+A is not here any more: Select all took it back, and the table above is where it is checked.
    // Nor is Ctrl+L, which the date sort took on 2026-09-23 beside Ctrl+J and Ctrl+K.
    var goneChords = [["1", ctrl], ["2", ctrl], ["3", ctrl], ["D", ctrl], ["U", ctrl],
                      ["E", ctrl], ["W", ctrl],
                      ["Tab", ctrl], ["Space", ctrl], ["PageUp", ctrl], ["PageDown", ctrl],
                      ["P", alt], ["Down", shift], ["Up", shift], ["Delete", shift],
                      ["Backspace", none], ["Home", none], ["End", none],
                      ["PageUp", none], ["PageDown", none], ["Q", alt], ["J", meta]]
    for (var c = 0; c < goneChords.length; c++)
        key(goneChords[c][0], "", goneChords[c][1], "")

    // The TUI keeps Insert for copy and paste, which is the one pair a terminal cannot deliver as
    // Ctrl+C and Ctrl+V, and the digits that pick a tab. Neither reaches the window.
    key("Insert", "", ctrl, "copy", "listing", "tui")
    key("Insert", "", shift, "paste", "listing", "tui")
    key("Insert", "", ctrl, "", "listing", "gui")
    key("1", "1", none, "tab1", "listing", "tui")
    key("9", "9", none, "tab9", "listing", "tui")
    key("1", "1", none, "", "listing", "gui")

    // The rail, the menus, the panels and the three preview kinds, which the listing table does not
    // reach at all: a shared row answers only the listing and the rail, so the rest is preset rows.
    var overlays = ["menu", "panel", "preview", "pdf", "media"]
    for (var o = 0; o < overlays.length; o++) {
        var context = overlays[o]
        key("Down", "", none, "cursorDown", context)
        key("Up", "", none, "cursorUp", context)
        key("Return", "", none, "open", context)
        key("Enter", "", none, "open", context)
        key("Space", " ", none, "preview", context)
        key("Escape", "", none, "escape", context)
        key("Tab", "", shift, "focusPrevious", context)
        // No letter reaches any of them now, and neither does a chord the listing lost.
        key("J", "j", none, "", context)
        key("K", "k", none, "", context)
        key("D", "d", none, "", context)
        key("X", "", ctrl, "", context)
        key("N", "", ctrl, "", context)
        key("Tab", "", ctrl, "", context)
    }
    key("Menu", "", none, "menu", "rail")
    key("Down", "", none, "cursorDown", "rail")
    key("Escape", "", none, "escape", "editor")
    key("Return", "", none, "", "editor")

    // The context menu steps on the arrows, and Right walks into a submenu where Left walks out.
    key("Right", "", none, "menuRight", "menu")
    key("Left", "", none, "parent", "menu")
    // The 2026-09-18 ruling, "all preview files behave the same way": one action per direction in
    // all three preview contexts, and ui/js/PreviewKeys.js act is the only thing that reads the
    // expanded state to decide what it moves. The old table gave Left three meanings across the
    // kinds, which is what a row per kind buys and what this asserts is gone.
    key("Left", "", none, "previewBack", "preview")
    key("Left", "", none, "previewBack", "pdf")
    key("Left", "", none, "previewBack", "media")
    key("Right", "", none, "previewForward", "preview")
    key("Right", "", none, "previewForward", "pdf")
    key("Right", "", none, "previewForward", "media")
    // And the two names that pair of rows used to carry are bound nowhere at all now: the whole
    // effective table is read for them rather than a context list being guessed at.
    check("no key spells a seek any more", Keymap.bindingRows("default", "gui")
          .filter(function (row) { return row.action === "seekBack" || row.action === "seekForward" })
          .length, 0)
    // MediaMute: m is the menu in the listing and the mute in a media preview, its two meanings.
    key("M", "m", none, "mute", "media")
    // The PDF's zoom and expand went with the bare letters; the strip's own controls are the route.
    key("Minus", "-", none, "", "pdf")
    key("Plus", "+", shift, "", "pdf")
    key("E", "e", none, "", "pdf")

    // The sheet advertises what is bound and nothing else, so it is read back off the same table.
    function capFor(action) {
        var listed = Keymap.sheetFor("default", "gui")
        for (var s = 0; s < listed.length; s++)
            if (listed[s].action === action) return listed[s].keys
        return ""
    }
    check("the sheet spells move with the arrow", capFor("cursorDown"), "down")
    check("the sheet spells parent with the arrow", capFor("parent"), "left")
    check("the sheet spells browse in with the arrow", capFor("pageForward"), "right")
    // Both halves of the ruling are on the card: the key that marks a row and the chord that marks
    // the folder, which is the pair a listing cannot be operated from the keyboard without.
    check("the sheet spells select with the key that does it", capFor("toggleSelect"), "space")
    check("and select all with its chord", capFor("selectAll"), "ctrl-a")
    check("and the reload with the key every file manager uses", capFor("refresh"), "f5")
    check("the sheet spells copy with its chord", capFor("copy"), "ctrl-c")
    check("the sheet spells the file path with its chord", capFor("copypath"), "ctrl-b")
    check("the sheet spells the sidebar with its chord", capFor("sidebar"), "ctrl-g")
    check("the sheet spells trash with the bare key", capFor("trash"), "delete")
    check("the sheet draws no row for a key that left", capFor("filter"), "")
    check("nor for the tab keys", capFor("tabNew"), "")

    Keymap.setPreset("unknown")
    check("an unknown stored preset resolves to Default", Keymap.preset, "default")
    // A menu row draws the key the operator presses. Every one of these is a chord now, and before
    // the hint rule took chords they all drew nothing while their rows went on claiming a key.
    check("Copy hints its chord", Keymap.hintFor("copy"), "ctrl-c")
    check("Paste hints its chord", Keymap.hintFor("paste"), "ctrl-v")
    check("Trash hints the bare key it answers to", Keymap.hintFor("trash"), "delete")
    check("Copy path hints its chord", Keymap.hintFor("copypath"), "ctrl-b")
    check("Rename hints its own key", Keymap.hintFor("rename"), "f2")
    check("menu-only actions invent no shortcut", Keymap.hintFor("emptyTrash"), "")
    check("and neither do the actions that lost their key", Keymap.hintFor("openTerminal"), "")

    check("the sheet is populated from effective current bindings", Keymap.SHEET.length > 15, true)
    var widestCap = 0, identifierLabel = ""
    var listed = Keymap.sheetFor("default", "gui")
    for (var r = 0; r < listed.length; r++) {
        if (listed[r].keys.length > widestCap) widestCap = listed[r].keys.length
        if (/[a-z][A-Z]/.test(listed[r].label)) identifierLabel = listed[r].label
        if (listed[r].keys.split(" / ").length > 2) identifierLabel = "too many spellings: " + listed[r].keys
    }
    check("no cap outgrows its half of the card", widestCap <= 18, true)
    check("no row prints an action id where its wording belongs", identifierLabel, "")

    // MediaMute rule 4: the preview's own mute key is listed once, under Look, while the listing
    // goes on advertising m as its menu key.
    var muteRows = 0, muteCap = ""
    for (var q = 0; q < listed.length; q++)
        if (listed[q].action === "mute") { muteRows++; muteCap = listed[q].keys }
    check("the sheet lists mute once", muteRows, 1)
    check("and lists it under its own key", muteCap, "m")
    check("while m still opens the menu in the listing", Keymap.lookupFor("default", 0, "m", 0, "listing", "gui"), "menu")
    check("the sheet group that claims it is Look", Keymap.SHEET_GROUPS.look.indexOf("mute") >= 0, true)
    // The split is what is drawn and not where the cursor goes, so it is claimed beside the rail.
    check("the split is claimed under Look too", Keymap.SHEET_GROUPS.look.indexOf("toggleDual") >= 0, true)

    check("pointer contract remains populated", Keymap.POINTER.length > 10, true)
    var effective = Keymap.bindingRows("default", "gui")
    check("every effective binding resolves to the action it advertises", effective.every(function (row) {
        return Keymap.lookupFor("default", row.keycode, row.text, row.mask, "listing", "gui") === row.action
    }), true)
    check("and the table is the size the strip left it, plus the split's chord, the creation trio, Select all, F5, the deep search and the three sort chords",
          effective.length, 32)
}

File: src/tui/keymap.rs
use super::input::Key;
use std::collections::HashMap;

#[derive(Default)]
pub struct Map {
    blocks: Vec<(String, HashMap<String, String>)>,
}
impl Map {
    pub fn load() -> Self {
        Self::parse(include_str!("../../keys.toml"))
    }
    // Sample input: [[text]]# comment, followed by char = "j" and action = "cursorDown".
    fn parse(text: &str) -> Self {
        let mut map = Self::default();
        for line in text.lines().map(str::trim) {
            if let Some((kind, suffix)) = line
                .strip_prefix("[[")
                .and_then(|header| header.split_once("]]"))
            {
                if suffix.trim().is_empty() || suffix.trim_start().starts_with('#') {
                    map.blocks.push((kind.trim().into(), HashMap::new()));
                    continue;
                }
            }
            if line.starts_with('#') {
                continue;
            }
            if let Some((key, value)) = line.split_once('=') {
                let value = value.trim();
                if let Ok(crate::jsondoc::Json::Str(value)) = crate::jsondoc::parse(value) {
                    if let Some((_, block)) = map.blocks.last_mut() {
                        block.insert(key.trim().into(), value);
                    }
                }
            }
        }
        map
    }
    pub fn action(&self, key: &Key, preset: &str) -> String {
        self.in_context(key, preset, "listing")
    }
    pub fn in_context(&self, key: &Key, preset: &str, context: &str) -> String {
        if context == "listing" && key.name == "Insert" && key.mods == "ctrl" {
            return "copy".into();
        }
        if context == "listing" && key.name == "Insert" && key.mods == "shift" {
            return "paste".into();
        }
        for name in [preset, "all"] {
            for (kind, block) in &self.blocks {
                let here = get(block, "context");
                let mods = get(block, "mods");
                let key_matches = if mods == "text" {
                    matches!(key.mods.as_str(), "" | "shift") && get(block, "key") == key.text
                } else {
                    (if mods == "none" { "" } else { mods }) == key.mods
                        && get(block, "key") == key.name
                };
                if kind == "preset"
                    && get(block, "name") == name
                    && get(block, "frontend") != "gui"
                    && (here.split(',').any(|part| part == context || part == "all")
                        || (here.is_empty() && context == "listing"))
                    && key_matches
                {
                    return get(block, "action").into();
                }
            }
        }
        if context != "listing" {
            return String::new();
        }
        for (kind, block) in &self.blocks {
            let matches = if key.mods.is_empty() {
                (kind == "text" && get(block, "char") == key.text && !key.text.is_empty())
                    || (kind == "code" && get(block, "key") == key.name)
            } else {
                kind == &key.mods && get(block, "key") == key.name
            };
            if matches {
                return get(block, "action").into();
            }
        }
        String::new()
    }
    pub fn sheet(&self, preset: &str) -> Vec<String> {
        let mut rows = Vec::new();
        for (_, label) in self.blocks.iter().filter(|(kind, _)| kind == "sheet") {
            let action = get(label, "action");
            if !supported(action) {
                continue;
            }
            let mut keys = Vec::new();
            for (kind, block) in &self.blocks {
                let mods = if kind == "preset" {
                    get(block, "mods")
                } else {
                    kind
                };
                let name = if kind == "text" {
                    get(block, "char")
                } else {
                    get(block, "key")
                };
                if name.is_empty()
                    || !matches!(
                        mods,
                        "text"
                            | "code"
                            | "none"
                            | ""
                            | "shift"
                            | "ctrl"
                            | "ctrlshift"
                            | "alt"
                            | "super"
                            | "supershift"
                            | "superalt"
                    )
                {
                    continue;
                }
                let text_key = mods == "text";
                let key = if text_key {
                    Key::character(name.chars().next().unwrap(), "")
                } else {
                    Key::named(
                        name,
                        if matches!(mods, "code" | "none") {
                            ""
                        } else {
                            mods
                        },
                    )
                };
                let found = self.action(&key, preset);
                if canonical(&found) != action {
                    continue;
                }
                let prefix = match key.mods.as_str() {
                    "ctrl" => "Ctrl+",
                    "ctrlshift" => "Ctrl+Shift+",
                    "alt" => "Alt+",
                    "super" => "Super+",
                    "supershift" => "Super+Shift+",
                    "superalt" => "Super+Alt+",
                    "shift" => "Shift+",
                    _ => "",
                };
                let chord = format!(
                    "{}{}{}",
                    prefix,
                    name,
                    if found.ends_with("Arm") { name } else { "" }
                );
                if !keys.contains(&chord) {
                    keys.push(chord);
                }
            }
            if !keys.is_empty() {
                rows.push(format!("{}  {}", keys.join(" / "), get(label, "label")));
            }
        }
        rows.push("1–9  Select tab".into());
        rows.push("q  Quit".into());
        rows
    }
}
fn canonical(action: &str) -> &str {
    match action {
        "copyArm" => "copy",
        "cutArm" => "cut",
        "pasteArm" => "paste",
        "cursorFirstArm" => "cursorFirst",
        "trashArm" => "trash",
        _ => action,
    }
}
fn supported(action: &str) -> bool {
    matches!(
        action,
        "cursorDown"
            | "cursorUp"
            | "extendDown"
            | "extendUp"
            | "pageDown"
            | "pageUp"
            | "cursorFirst"
            | "cursorLast"
            | "parent"
            | "historyBack"
            | "historyForward"
            | "open"
            | "toggleSelect"
            | "selectAll"
            | "toggleHidden"
            | "pathBar"
            | "filter"
            | "search"
            | "rename"
            | "newFolder"
            | "newFile"
            | "properties"
            | "copy"
            | "cut"
            | "paste"
            | "movePaste"
            | "duplicate"
            | "trash"
            | "deletePermanently"
            | "undo"
            | "redo"
            | "sortNext"
            | "sortReverse"
            | "tabNew"
            | "openTerminal"
            | "windowNew"
            | "tabClose"
            | "tabNext"
            | "tabPrevious"
            | "preview"
            | "togglePreview"
            | "loadPreview"
            | "focusPreview"
            | "focusNext"
            | "menu"
            | "keymapSheet"
            | "reveal"
            | "escape"
            | "quit"
    )
}
fn get<'a>(map: &'a HashMap<String, String>, key: &str) -> &'a str {
    map.get(key).map(String::as_str).unwrap_or("")
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn compiled_source_drives_text_and_native_ctrl() {
        let map = Map::load();
        // The shared [[text]] table, which m is the one letter left in.
        assert_eq!(
            map.action(
                &Key {
                    name: "M".into(),
                    text: "m".into(),
                    mods: "".into(),
                    pointer: None,
                },
                "default"
            ),
            "menu"
        );
        // A letter the table no longer carries at all: navigation is the four arrows.
        assert_eq!(
            map.action(
                &Key {
                    name: "J".into(),
                    text: "j".into(),
                    mods: "".into(),
                    pointer: None,
                },
                "default"
            ),
            ""
        );
        assert_eq!(
            map.action(
                &Key {
                    name: "Down".into(),
                    text: "".into(),
                    mods: "".into(),
                    pointer: None,
                },
                "default"
            ),
            "cursorDown"
        );
        assert_eq!(
            map.action(
                &Key {
                    name: "Insert".into(),
                    text: "".into(),
                    mods: "shift".into(),
                    pointer: None,
                },
                "default"
            ),
            "paste"
        );
    }
    #[test]
    fn contexts_share_only_declared_actions_and_sheet_matches_preset() {
        let map = Map::load();
        assert_eq!(map.in_context(&Key::character('d', ""), "default", "pdf"), "");
        assert_eq!(
            map.in_context(&Key::named("Space", ""), "default", "pdf"),
            "preview"
        );
        assert_eq!(
            map.in_context(&Key::named("Tab", "shift"), "default", "media"),
            "focusPrevious"
        );
        assert_eq!(
            map.in_context(&Key::named("Insert", "ctrl"), "default", "editor"),
            ""
        );
        assert!(map
            .sheet("default")
            .iter()
            .any(|line| line.starts_with("Ctrl+C") && line.ends_with("  copy")));
        assert!(!map
            .sheet("default")
            .iter()
            .any(|line| line.to_lowercase().contains("grid")));
    }
    #[test]
    fn header_comments_do_not_strip_quoted_hashes_or_accept_other_suffixes() {
        let map = Map::parse(r##"
[[text]] # A header comment can contain = and ]].
char = "#"
action = "literal#action"
# [[preset]] is still a comment.
[[text]]# No intervening space is required.
char = "j"
action = "cursorDown"
[[code]] unexpected
"##);
        assert_eq!(map.blocks.len(), 2);
        assert_eq!(map.action(&Key::character('#', ""), "default"), "literal#action");
        assert_eq!(map.action(&Key::character('j', ""), "default"), "cursorDown");
    }
}
